docs/content/sips/012-signing-spin-releases.md

```diff
@@ -0,0 +1,206 @@
+
+## Proposal
+
+This SIP proposes that for every [Spin release](https://github.com/fermyon/spin/releases),
+we add metadata that helps a user validate the integrity of the package they
+download.
+Specifically, this proposal suggests the use of [Sigstore](https://www.sigstore.dev/),
+_a new standard for signing, verifying, and protecting software_, and in particular,
+the use of [the new release of Cosign v2.0](https://blog.sigstore.dev/cosign-2-0-released),
+which stabilizes [keyless signatures using an OIDC provider](https://docs.sigstore.dev/cosign/keyless/).
+
+In keyless mode, Sigstore makes an OIDC provider the root of trust for the
+signature by creating short-lived x509 certificates bound to an OIDC identity.
+The certificates are authenticated and auditable, which makes the resulting
+signatures auditable as well.
+
+This process uses a few central pieces that are assumed trusted:
+
+- an OIDC provider (such as GitHub)
+- Fulcio — free root certificate authority that issues temporary certificates
+bound to an OIDC identity, published to Rekor
+- Rekor — transparency and timestamp service, provides a ledger that can be
+audited
+- `cosign sign` — the CLI that connects an OIDC identity to Fulcio to generate the
+certificate, then uses the certificate to sign the artifact, publishing the
+signing certificate to Rekor
+- `cosign verify` — the CLI that verifies the signature of a given artifact by
+auditing the transparency log
+
+![Sigstore workflow, as described by sigstore.dev](../../static/image/sigstore.png)
+
+This document proposes that the build and release infrastructure for the Spin project be the entity
+that signs the release artifacts before finalizing a new release using
+[GitHub's OIDC workflow](https://docs.github.com/en/actions/deployment/security-hardening-your-deployments/about-security-hardening-with-openid-connect).
+This would tell a user that the artifact they are downloading was built by the
+Spin project GitHub infrastructure, and that it has not been tampered with since
+its creation.
+
+The following workflow describes the process:
+
+- `permission: id-token: write` needs to be added to the GitHub action creating
+the release (see [permissions for `GITHUB_TOKEN`](https://docs.github.com/en/actions/security-guides/automatic-token-authentication#permissions-for-the-github_token))
+- a new step is added in the Spin release process that signs every artifact and
+captures the certificate and signature:
+
+```bash
+$ cosign sign-blob \
+    # write the output certificate to a file that will be later added to the release
+    --output-certificate out/spin-v1.0.0-linux-amd64-keyless.pem \
+    # write the signature to a file that will be later added to the release
+    --output-signature out/spin-v1.0.0-linux-amd64-keyless.sig spin- \
+    # skip interactive confirmation
+    --yes
+```
+
+- when uploading the assets to the new release, the certificate and signatures
+need to be added side-by-side with the actual artifact
```

Comment:
It depends on what we want to sign — the archive containing everything, or just the binary. 

Since we already compute the content digest _of the archive_, it might make sense to sign that. If we sign the archive, then the signature and certificate would be distributed as separate release artifacts.

If we only sign the binary, then we could distribute the signature and certificate inside the archive.